### PR TITLE
Embedded field and small refactoring of documents

### PR DESCRIPTION
**Pull request by radeklos**
Embedded document type.. 
- [x] Docs

### CODE REVIEW

File: flask_hal/document.py [hunk 1]
```diff
@@ -38,19 +38,18 @@ def __init__(self, data=None, links=None, embedded=None):
...
+    @property
+    def links(self):
+        return self._links
 
-        self.links = links
+    @links.setter
+    def links(self, value):
+        if not isinstance(value, link.Collection):
+            raise TypeError('links must be a {} instance'.format(link.Collection))
+        self._links = value
```
Comment (krak3n) [inline flask_hal/document.py:52]: :+1:

**Comment by krak3n** [comment]
This like a nice simple solution.

Would the usage be something like this:

``` python
objs = Order.query.all()
total = Order.query.count()
# assuming a links function exists on the Order class which generates a links Collection
orders = [Embedded(Serializer().serialize(o), links=o.links()) for o in objs] 

return Document(data={'total': total}, embedded=orders)
```

**Comment by radeklos** [comment]
You shouldn't pass a list as embedded:

``` python
Document(data={'total': total}, embedded={'orders': orders})
```

**Comment by krak3n** [comment]
Ah ofc, cool :+1:

**Comment by radeklos** [comment]
Or it could be

``` python
orders = [Embedded('orders', Serializer().serialize(o), links=o.links()) for o in obks] 
```

But then we have to manage `list` -> `dict` convention and hybrid `embedded` setter..

File: flask_hal/document.py [hunk 2]
```diff
@@ -82,3 +95,65 @@ def to_json(self):
...
+        >>>     data={'currentlyProcessing': 14}
+        >>> )
+        >>> document.to_json()
+        ... {
+                "_links": {
+                    "self": {"href": "http://localhost/entity/231"}
+                },
+                "_embedded": {
+                    "orders": {
+                        "_embedded": {
```
Comment (krak3n) [inline flask_hal/document.py:146]: Should be `_links`?

Comment (radeklos) [inline flask_hal/document.py:146]: Sorry don't get it.. Links are on line 150. What do you means? And `_links` are options so when it's empty document doesn't return them.

Comment (krak3n) [inline flask_hal/document.py:146]: Ah, my bad, totally missed the `_links` underneath this, basically we nest forever:

``` json
{
    "_links": {
        "self": { "href": "/orders" },
    },
    "_embedded": {
        "orders": [{
            "_links": {
                "self": { "href": "/orders/123" },
                "basket": { "href": "/baskets/98712" },
                "customer": { "href": "/customers/7809" }
            },
            "_embedded": {
                "foo": {
                    "name": "foo"
                }
            },
            "total": 30.00,
            "currency": "USD",
            "status": "shipped"
            },{
            "_links": {
                "self": { "href": "/orders/124" },
                "basket": { "href": "/baskets/97213" },
                "customer": { "href": "/customers/12369" }
            },
            "_embedded": {
                "bar": {
                    "name": "bar"
                }
            },
            "total": 20.00,
            "currency": "USD",
            "status": "processing"
        }]
    },
    "currentlyProcessing": 14,
    "shippedToday": 20
}
```

**Comment by krak3n** [comment]
I say merge this :bomb:!

<sup>then write some tests</sup>

File: flask_hal/document.py (hunk 2 above)
Comment (radeklos) [inline flask_hal/document.py:146]: :+1:

**Pull request merged**